<h2>vapply</h2>
Như thường lệ, chúng ta sẽ quan sát tổng quan tập dữ liệu flags.

In [1]:
flags <- read.csv(url("http://archive.ics.uci.edu/ml/machine-learning-databases/flags/flag.data"))
# Set column names
colnames(flags) <- c("name", "landmass", "zone", "area", "population",
                     "language", "religion", "bars", "stripes", "colours",
                     "red", "green", "blue", "gold", "white", "black",
                     "orange", "mainhue", "circles", "crosses", "saltires",
                     "quarters", "sunstars", "crescent", "triangle",
                     "icon", "animate", "text", "topleft", "botright")

# liệt kê 6 dòng dữ liệu ban đầu
head(flags)

name           landmass zone area population language religion bars stripes
1 Albania        3        1      29  3          6       6        0    0      
2 Algeria        4        1    2388 20          8       2        2    0      
3 American-Samoa 6        3       0  0          1       1        0    0      
4 Andorra        3        1       0  0          6       0        3    0      
5 Angola         4        2    1247  7         10       5        0    2      
6 Anguilla       1        4       0  0          1       1        0    1      
  colours ⋯ saltires quarters sunstars crescent triangle icon animate text
1 3       ⋯ 0        0        1        0        0        0    1       0   
2 3       ⋯ 0        0        1        1        0        0    0       0   
3 5       ⋯ 0        0        0        0        1        1    1       0   
4 3       ⋯ 0        0        0        0        0        0    0       0   
5 3       ⋯ 0        0        1        0        0        1    0       0   
6 3       ⋯ 0        0        0        0        0        0    1       0   
  topleft botright
1 red     red     
2 green   white   
3 blue    red     
4 blue    red     
5 red     black   
6 white   blue

In [2]:
# quan sát số chiều của tập dữ liệu
dim(flags)

[1] 193  30

In [3]:
# quan sát cấu trúc của tập dữ liệu
str(flags)

'data.frame':	193 obs. of  30 variables:
 $ name      : Factor w/ 193 levels "Albania","Algeria",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ landmass  : int  3 4 6 3 4 1 1 2 2 6 ...
 $ zone      : int  1 1 3 1 2 4 4 3 3 2 ...
 $ area      : int  29 2388 0 0 1247 0 0 2777 2777 7690 ...
 $ population: int  3 20 0 0 7 0 0 28 28 15 ...
 $ language  : int  6 8 1 6 10 1 1 2 2 1 ...
 $ religion  : int  6 2 1 0 5 1 1 0 0 1 ...
 $ bars      : int  0 2 0 3 0 0 0 0 0 0 ...
 $ stripes   : int  0 0 0 0 2 1 1 3 3 0 ...
 $ colours   : int  3 3 5 3 3 3 5 2 3 3 ...
 $ red       : int  1 1 1 1 1 0 1 0 0 1 ...
 $ green     : int  0 1 0 0 0 0 0 0 0 0 ...
 $ blue      : int  0 0 1 1 0 1 1 1 1 1 ...
 $ gold      : int  1 0 1 1 1 0 1 0 1 0 ...
 $ white     : int  0 1 1 0 0 1 1 1 1 1 ...
 $ black     : int  1 0 0 0 1 0 1 0 0 0 ...
 $ orange    : int  0 0 1 0 0 1 0 0 0 0 ...
 $ mainhue   : Factor w/ 8 levels "black","blue",..: 7 5 2 4 7 8 7 2 2 2 ...
 $ circles   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ crosses   : int  0 0 0 0 

In [4]:
# tổng hợp thông tin về tập dữ liệu
summary(flags)

             name        landmass          zone            area        
 Albania       :  1   Min.   :1.000   Min.   :1.000   Min.   :    0.0  
 Algeria       :  1   1st Qu.:3.000   1st Qu.:1.000   1st Qu.:    9.0  
 American-Samoa:  1   Median :4.000   Median :2.000   Median :  111.0  
 Andorra       :  1   Mean   :3.565   Mean   :2.218   Mean   :  700.3  
 Angola        :  1   3rd Qu.:5.000   3rd Qu.:4.000   3rd Qu.:  463.0  
 Anguilla      :  1   Max.   :6.000   Max.   :4.000   Max.   :22402.0  
 (Other)       :187                                                    
   population         language         religion          bars      
 Min.   :   0.00   Min.   : 1.000   Min.   :0.000   Min.   :0.000  
 1st Qu.:   0.00   1st Qu.: 2.000   1st Qu.:1.000   1st Qu.:0.000  
 Median :   4.00   Median : 6.000   Median :1.000   Median :0.000  
 Mean   :  23.31   Mean   : 5.316   Mean   :2.192   Mean   :0.456  
 3rd Qu.:  14.00   3rd Qu.: 9.000   3rd Qu.:4.000   3rd Qu.:0.000  
 Max.   :1008.00

Như ở phần trước, hàm unique() trả về một vector chứa các giá trị đơn nhất từ đối tượng truyền vào. Do đó, sapply(flags, unique) sẽ trả về list một vector chứa các giá trị đơn nhất cho từng cột dữ liệu của flags.

Khi làm việc trong môi trường tương tác (interactive at the prompt), chúng ta sẽ ít khi bị nhầm lẫn vì bạn có thể nhìn thấy kết quả ngay tức thì, từ đó có thể chỉnh sửa lại tính toán của mình. Tuy nhiên, khi làm việc trên môi trường không tương tác (non-interactively: khi viết functions), chúng ta dễ dự đoán sai kết quả trả về. Vì vậy, chúng ta cần cảnh giác hơn bằng cách sử dụng hàm vapply().

Trong khi sapply() cố gắng tinh giản kết quả trả về thì vapply() cho phép bạn lựa chọn định dạng kết quả trả về một cách tường minh. Nếu như kết quả trả về không khớp với định dạng mà bạn chọn, vapply() sẽ đưa ra thông báo lỗi và dừng hẳn chương trình. Việc làm này có thể ngăn chặn được các vấn đề quan trọng trong code của bạn khi bạn dự đoán sai kết quả trả về do nhầm lẫn.

Chúng ta sẽ thử dòng lệnh vapply(flags, unique, numeric(1)). Dòng lệnh này có nghĩa là bạn mong muốn mỗi phần tử trong kết quả trả về là numeric vector có độ dài là 1. Trong khi điều này không đúng, bạn sẽ nhận một thông báo lỗi.


In [5]:
vapply(flags, unique, numeric(1))

ERROR: Error in vapply(flags, unique, numeric(1)): values must be length 1,
 but FUN(X[[1]]) result is length 193


Sử dụng lại hàm sapply(flags, class) để trả về một character vector chứa tên kiểu dữ liệu cho các cột thuộc tính trong tập dữ liệu. Giả sử chúng ta muốn lựa chọn định dạng trả về như sapply() bằng cách sử dụng hàm vapply() ta sẽ làm như sau.


In [6]:
print(sapply(flags, class))

      name   landmass       zone       area population   language   religion 
  "factor"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
      bars    stripes    colours        red      green       blue       gold 
 "integer"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
     white      black     orange    mainhue    circles    crosses   saltires 
 "integer"  "integer"  "integer"   "factor"  "integer"  "integer"  "integer" 
  quarters   sunstars   crescent   triangle       icon    animate       text 
 "integer"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
   topleft   botright 
  "factor"   "factor" 


In [7]:
print(vapply(flags, class, character(1)))

      name   landmass       zone       area population   language   religion 
  "factor"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
      bars    stripes    colours        red      green       blue       gold 
 "integer"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
     white      black     orange    mainhue    circles    crosses   saltires 
 "integer"  "integer"  "integer"   "factor"  "integer"  "integer"  "integer" 
  quarters   sunstars   crescent   triangle       icon    animate       text 
 "integer"  "integer"  "integer"  "integer"  "integer"  "integer"  "integer" 
   topleft   botright 
  "factor"   "factor" 


Đối số 'character(1)' mách cho R biết rằng chúng ta mong muốn hàm class() trả về một character vector có độ dài là 1 khi áp dụng lên từng cột thuộc tính của Flags. Dĩ nhiên, kết quả trả về tương tự như hàm sapply().

Bạn có thể nghĩ rằng vapply() sẽ an toàn hơn sapply(), khi nó đòi hỏi bạn phải chỉ định kết quả trả về, thay vì để R cố gắng tinh giản kết quả cho bạn. Thêm vào đó, vapply() có tốc độ xử lý nhanh hơn sapply() trên tập dữ liệu lớn. Tuy nhiên, khi phân tích dữ liệu theo kiểu tương tác (interactively at the prompt), sapply() sẽ tiết kiệm thời gian trong việc nhập lệnh.

<h2>tapply</h2>
Khi phân tích dữ liệu, bạn muốn tách dữ liệu của mình thành các nhóm dựa trên giá trị của một vài cột thuộc tính. Sau đó, ta áp dụng một function lên các thành viên trong nhóm này. Hàm tiếp theo chúng ta sẽ khảo sát đó là tapply(), sẽ hỗ trợ chúng ta làm điều này.

Cột thuộc tính 'landmass' trong tập dữ liệu gồm các giá trị integer nằm trong khoảng 1 và 6, mỗi giá trị thể hiện một phần khác nhau trên bản đồ thế giới. Sử dụng hàm table() để xem có bao nhiêu flags/countries nằm trong từng nhóm trên.

In [8]:
# similar to histogram
table(flags$landmass)


 1  2  3  4  5  6 
31 17 35 52 38 20 

Cột thuộc tính 'animate" trong tập dữ liệu chứa giá trị 1 nếu quốc kì của quốc gia đó chứa một biểu tượng animate (ví dụ như một con chim ưng, một cái cây, một bàn tay) và 0 ngược lại. Sử dụng hàm table() để xem có bao nhiêu quốc kì chứa các biểu tượng này.

In [9]:
table(flags$animate)


  0   1 
154  39 

Kết quả trên cho chúng ta biết có 39 quốc kì chứa các biểu tượng animate (animate = 1) và 155 không có animate (animate = 0).

Sử dụng hàm ```tapply(flags$animate, flags$landmass, mean)``` để tính giá trị trung bình cho cột thuộc tính 'animate' ứng với từng châu lục 'landmass'. Khi đó, ta sẽ có số lượng trung bình của các quốc kì có biểu tượng 'animate' nằm trong từng châu lục.


In [10]:
args(tapply)

function (X, INDEX, FUN = NULL, ..., default = NA, simplify = TRUE) 
NULL

In [11]:
print(tapply(flags$animate, flags$landmass, mean))

        1         2         3         4         5         6 
0.4193548 0.1764706 0.1142857 0.1346154 0.1578947 0.3000000 


Châu lục đầu tiên (landmass = 1) thuộc North America có tỉ lệ cao nhất so với các châu lục còn lại về số quốc kì có biểu tượng animate (0.4194).

Trong bài viết này, chúng ta đã học về cách sử dụng hàm vapply() để có kết quả trả về an toàn hơn thay vì sử dụng hàm sapply(), hàm này đặc biệt hữu ích khi bạn viết functions của riêng mình. Đồng thời, chúng ta cũng đã học cách sử dụng hàm tapply() để tách dữ liệu của chúng ta thành từng nhóm dựa trên giá trị của một vài cột thuộc tính, sau đó áp dụng một hàm tính toán nào đó lên từng nhóm này. Các hàm này sẽ hỗ trợ bạn thao tác tốt hơn trong quá trình phân tích dữ liệu.

<strong>Nguồn tham khảo:</strong> <a href="http://swirlstats.com/" target="_blank" rel="noopener">http://swirlstats.com/</a>